### PR TITLE
Fix breadcrumb animation

### PR DESCRIPTION
**Pull request by Exalm**
* Use add_tick_callback() instead of counting frames manually
* Use easeOutCubic interpolator
* Don't clip the arrow at the end to avoid blinking
* Don't stop the animation until the element is fully hidden, i.e. animate it going under the arrow of the previous element as well.
* Align arrow position to pixel grid.

Peek 2019-10-06 23-45_1

It still blinks when changing location though, but that's a different issue.

### CODE REVIEW

File: libwidgets/Chrome/BreadcrumbElement.vala [hunk 1]
```diff
@@ -197,6 +198,10 @@ public class Marlin.View.Chrome.BreadcrumbElement : Object {
             icon_to_draw = PF.PixbufUtils.lucent (icon_to_draw, 50);
         }
 
+        if (offset_px >= width) {
+            room_for_text = false;
```
Comment (Exalm) [inline libwidgets/Chrome/BreadcrumbElement.vala]: Hm, and this is not correct, I've missed icon here.

Comment (Exalm) [reply libwidgets/Chrome/BreadcrumbElement.vala]: I've redone this with masking. See commit descriptions for details, there's sth weird going on with clipping + `render_layout()`.

**Comment by Exalm** [comment]
Sorry for the churn, found a few code style issues. :)

File: libwidgets/Chrome/BasicBreadcrumbsEntry.vala [hunk 2]
```diff
@@ -584,33 +584,35 @@ namespace Marlin.View.Chrome {
...
+            var anim = add_tick_callback ((widget, frame_clock) => {
+                int64 time = frame_clock.get_frame_time ();
+                double t = (double) (time - start_time) / LOCATION_BAR_ANIMATION_TIME_USEC;
+                t = 1 - ease_out_cubic (t.clamp (0, 1));
+
+                anim_state = final_offset + (initial_offset - final_offset) * t;
 
+                foreach (BreadcrumbElement bread in els) {
+                    bread.offset = anim_state;
+                }
```
Comment (jeremypw) [inline libwidgets/Chrome/BasicBreadcrumbsEntry.vala]: Lint: blank line after loop clause

File: libwidgets/Chrome/BreadcrumbElement.vala [hunk 3]
```diff
@@ -79,6 +79,23 @@ public class Marlin.View.Chrome.BreadcrumbElement : Object {
         this.icon_info = icon_info;
     }
 
+    private Cairo.Surface? get_mask (double x1, double y1, double x2, double y2, int scale, Cairo.Path? clip_path) {
+        if (clip_path == null)
+            return null;
+
```
Comment (jeremypw) [inline libwidgets/Chrome/BreadcrumbElement.vala]: Lint: missing braces

File: libwidgets/Chrome/BreadcrumbElement.vala [hunk 4]
```diff
@@ -250,6 +279,15 @@ public class Marlin.View.Chrome.BreadcrumbElement : Object {
...
         }
 
+        var group = cr.pop_group ();
+        cr.set_source (group);
+        if (mask != null) {
+            cr.mask_surface (mask, clip_x1, clip_y1);
+        } else {
+            cr.paint ();
+        }
+        cr.restore ();
```
Comment (jeremypw) [inline libwidgets/Chrome/BreadcrumbElement.vala]: Lint: Missing blank line after if clause

**Comment by jeremypw** [comment]
The difference from master is quite hard to see on my system, I am afraid.  The redrawing of the unchanged breadcrumbs causes a slight flicker even when only adding or removing the final directory.  It may be that master is worse on other systems where the frame-rate is not 60Hz?  Nevertheless using add_tick_callback() is an improvement.

**Comment by jeremypw** [comment]
It probably needs a more extensive change to this widget e.g. by using an Overlay or Revealer and separating the BreadCrumbs from the Entry.

**Comment by Exalm** [comment]
> The difference from master is quite hard to see on my system, I am afraid. The redrawing of the unchanged breadcrumbs causes a slight flicker even when only adding or removing the final directory. 

This only touches the appear/disappear animation. I have elementary OS in a VM, and the difference is quite noticeable tbh. :)

Peek 2019-10-07 21-11

The flickering is still there indeed.

**Comment by Exalm** [comment]
The other flicker seems to be caused by the fact the pathbar briefly becomes insensitive, so it's not really drawing-related.

**Comment by Exalm** [comment]
Hm, I've resolved all discussions, why does github still say there are changes requested?

**Comment by jeremypw** [comment]
OK, maybe I am looking at the wrong thing.  You need to dismiss the previous review and request a new one (if the reviewer does not notice your changes). Its best to keep a separate commit when addressing review issues rather than force push so the reviewer can see what has changed.

**Comment by Exalm** [comment]
Ah, I see. That's quite different from the workflow I'm used to (GitLab doesn't have reviews, only separate discussions, so you just resolve them + GNOME usually doesn't squash commits so you correct them and force-push because they will be merged as is, so "fix this and that" is unwanted), but makes sense when the commits are squashed :) Will know, thanks!

**Pull request merged**